### PR TITLE
Fix concurrent insertions for Iceberg tables

### PR DESCRIPTION
**Pull request by pratyakshsharma**
## Description
Number of successful concurrent inserts/updates to an iceberg table is controlled by the table property `commit.retry.num-retries` which is by default set to 4 in TableProperties.java class in Iceberg. This PR aims to make this configurable.

## Motivation and Context
https://github.com/prestodb/presto/issues/21251

## Impact
Introduction of new user-facing config for iceberg connector: `iceberg.commit-retries`

## Test Plan
Manual and unit tests

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

Iceberg Changes
* Add the support to set commit.retry.num-retries table property with table creation to make the number of attempts to make in case of concurrent upserts configurable.
```

### CODE REVIEW

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConfig.java [hunk 1]
```diff
@@ -196,4 +198,17 @@ public HiveStatisticsMergeStrategy getHiveStatisticsMergeStrategy()
...
+    {
+        this.concurrentCommitRetries = concurrentRetries;
+        return this;
+    }
+
+    public int getConcurrentCommitRetries()
+    {
+        return concurrentCommitRetries;
+    }
 }
```
Comment (agrawalreetika) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConfig.java:214]: Since this is a table property for Iceberg to control the retry limit, would it be better to expose this as a table property instead similar to existing Iceberg table properties, WDYT? - https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/connector/iceberg.rst#table-properties

Comment (pratyakshsharma) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConfig.java:214]: makes sense. Will take care.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java [hunk 2]
```diff
@@ -283,9 +285,11 @@ public ConnectorOutputTableHandle beginCreateTable(ConnectorSession session, Con
             throw new TableAlreadyExistsException(schemaTableName);
         }
 
-        ImmutableMap.Builder<String, String> propertiesBuilder = ImmutableMap.builderWithExpectedSize(3);
+        ImmutableMap.Builder<String, String> propertiesBuilder = ImmutableMap.builderWithExpectedSize(4);
         FileFormat fileFormat = getFileFormat(tableMetadata.getProperties());
+        Integer commitRetries = getConcurrencyRetries(tableMetadata.getProperties());
         propertiesBuilder.put(DEFAULT_FILE_FORMAT, fileFormat.toString());
+        propertiesBuilder.put(COMMIT_NUM_RETRIES, String.valueOf(commitRetries));
```
Comment (agrawalreetika) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java:292]: Since this is commit retry property for Iceberg, then I think this should be applicable even for Iceberg tables that are created using the `Hadoop` or `Nessie` catalog?

**Comment by agrawalreetika** [comment]
@pratyakshsharma Could you please squash the commits and update Iceberg documentation as well for this?

**Review by agrawalreetika — CHANGES_REQUESTED** [review]

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java (hunk 2 above)
Comment (pratyakshsharma) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java:292]: yes it should be.

Comment (agrawalreetika) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java:292]: Then let's introduce this property in https://github.com/prestodb/presto/blob/master/presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergNativeMetadata.java#L157
Since the Table properties both both Metadata classes are the same, we can extract the common part for setting table properties and use it in both

**Review by agrawalreetika — CHANGES_REQUESTED** [review]

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java (hunk 2 above)
Comment (agrawalreetika) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java:292]: Then let's introduce this property in https://github.com/prestodb/presto/blob/master/presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergNativeMetadata.java#L157
Since the Table properties both both Metadata classes are the same, we can extract the common part for setting table properties and use it in both

Comment (pratyakshsharma) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java:292]: Sure. I will perform simple tests with nessie to verify this works, and then will introduce it here.

**Comment by pratyakshsharma** [comment]
the worker seems to crash as part of the failures in `linux-presto-e2e-tests`. 
cc @tdcmeehan @wanglinsong

**Comment by tdcmeehan** [comment]
> the worker seems to crash as part of the failures in `linux-presto-e2e-tests`. This happened second time consecutively in this PR. cc @tdcmeehan @wanglinsong

Can you please ask about this in the #presto-cpp channel in Slack?

**Review by steveburnett — APPROVED** [review]
LGTM! (docs)

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java (hunk 2 above)
Comment (pratyakshsharma) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java]: Nessie throws below exception when concurrent queries exceed the table config value - 
```
org.apache.iceberg.exceptions.CommitFailedException: Cannot commit: Reference hash is out of date. Update the reference 'main' and try again
	at org.apache.iceberg.nessie.NessieTableOperations.doCommit(NessieTableOperations.java:167)
	at org.apache.iceberg.BaseMetastoreTableOperations.commit(BaseMetastoreTableOperations.java:135)
	at org.apache.iceberg.BaseTransaction.lambda$commitSimpleTransaction$5(BaseTransaction.java:422)
	at org.apache.iceberg.util.Tasks$Builder.runTaskWithRetry(Tasks.java:413)
	at org.apache.iceberg.util.Tasks$Builder.runSingleThreaded(Tasks.java:219)
	at org.apache.iceberg.util.Tasks$Builder.run(Tasks.java:203)
	at org.apache.iceberg.util.Tasks$Builder.run(Tasks.java:196)
	at org.apache.iceberg.BaseTransaction.commitSimpleTransaction(BaseTransaction.java:418)
	at org.apache.iceberg.BaseTransaction.commitTransaction(BaseTransaction.java:302)
	at com.facebook.presto.iceberg.IcebergAbstractMetadata.finishInsert(IcebergAbstractMetadata.java:278)
	at com.facebook.presto.spi.connector.classloader.ClassLoaderSafeConnectorMetadata.finishInsert(ClassLoaderSafeConnectorMetadata.java:452)
	at com.facebook.presto.metadata.MetadataManager.finishInsert(MetadataManager.java:858)
	at com.facebook.presto.sql.planner.LocalExecutionPlanner.lambda$createTableFinisher$3(LocalExecutionPlanner.java:3392)
	at com.facebook.presto.operator.TableFinishOperator.getOutput(TableFinishOperator.java:289)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:428)
	at com.facebook.presto.operator.Driver.lambda$processFor$9(Driver.java:311)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:732)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:304)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1079)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:165)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:621)
	at com.facebook.presto.$gen.Presto_null__testversion____20231108_094807_1.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: org.projectnessie.error.NessieReferenceConflictException: Values of existing and expected content for key 'iceberg_db.test_nessie_iceberg' are different.
	at org.projectnessie.error.ErrorCode.lambda$asException$1(ErrorCode.java:66)
	at java.util.Optional.map(Optional.java:215)
	at org.projectnessie.error.ErrorCode.asException(ErrorCode.java:66)
	at org.projectnessie.client.rest.ResponseCheckFilter.checkResponse(ResponseCheckFilter.java:67)
	at org.projectnessie.client.rest.NessieHttpResponseFilter.filter(NessieHttpResponseFilter.java:29)
	at org.projectnessie.client.http.impl.jdk8.UrlConnectionRequest.lambda$executeRequest$2(UrlConnectionRequest.java:86)
	at java.util.Collections$SingletonList.forEach(Collections.java:4824)
	at org.projectnessie.client.http.impl.jdk8.UrlConnectionRequest.executeRequest(UrlConnectionRequest.java:86)
	at org.projectnessie.client.http.impl.jdk8.UrlConnectionRequest.post(UrlConnectionRequest.java:120)
	at org.projectnessie.client.http.HttpTreeClient.commitMultipleOperations(HttpTreeClient.java:202)
	at sun.reflect.GeneratedMethodAccessor1216.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.projectnessie.client.http.NessieHttpClient$ExceptionRewriter.invoke(NessieHttpClient.java:79)
	at com.sun.proxy.$Proxy522.commitMultipleOperations(Unknown Source)
	at org.projectnessie.client.http.v1api.HttpCommitMultipleOperations.commit(HttpCommitMultipleOperations.java:35)
	at org.apache.iceberg.nessie.NessieIcebergClient.commitTable(NessieIcebergClient.java:479)
	at org.apache.iceberg.nessie.NessieTableOperations.doCommit(NessieTableOperations.java:160)
	... 24 more
```

**Comment by agrawalreetika** [comment]
@pratyakshsharma Thanks for verifying it for Nessie as well.
Few things, 
1. Let's modify the Tests in https://github.com/prestodb/presto/blob/master/presto-iceberg/src/test/java/com/facebook/presto/iceberg/hive/TestIcebergSmokeHive.java#L59 with concurrency > 5 (Since 5 is the success case) and add for both the failure and success scenarios there. And if possible we can add similar tests for Nessie as well
2. I think we can modify the commit message to something like "Introduce iceberg table property `commit_retries` to handle concurrent insertions"?

**Comment by pratyakshsharma** [comment]
@agrawalreetika yes already working on fixing test cases. Sure will update the commit message.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java (hunk 2 above)
Comment (yingsu00) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java]: > Nessie throws below exception when concurrent queries exceed the table config value 

Does it work if concurrent queries not exceeding the table config value?

Comment (pratyakshsharma) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java]: yes, have added a test case for the same.

**Comment by pratyakshsharma** [comment]
I have commented the test case covering failure scenario for concurrent insertions, since it is failing and I am working on fixing it. I have added the expectedExceptions tag, but the test is failing despite that exception getting thrown.

**Comment by pratyakshsharma** [comment]
@yingsu00 As per our internal discussions, I checked JMeter codebase where they spawn new threads for load testing. They use `ThreadPoolExecutor` for managing the test plan, which in turn starts the new threads as per the configurations provided. `ThreadPoolExecutor` is just an implementation of ExecutorService. I tried following the same approach for my test case, but surprisingly it did not help. 
We can get on a quick call to discuss more.

File: presto-iceberg/src/test/java/com/facebook/presto/iceberg/nessie/TestIcebergSmokeNessie.java [hunk 3]
```diff
@@ -111,4 +122,68 @@ protected Table getIcebergTable(ConnectorSession session, String schema, String
...
+    {
+        final Session session = getSession();
+        assertUpdate(session, "CREATE TABLE test_concurrent_insert (col0 INTEGER, col1 VARCHAR) WITH (format = 'ORC', commit_retries = " + concurrency + ")");
+
+        int commitRetries = concurrency + 1;
+        final String[] strings = {"one", "two", "three", "four", "five", "six", "seven"};
+        final CountDownLatch countDownLatch = new CountDownLatch(commitRetries);
+        AtomicInteger value = new AtomicInteger(0);
+        Set<Throwable> errors = new CopyOnWriteArraySet<>();
+        List<Thread> threads = Stream.generate(() -> new Thread(() -> {
```
Comment (tdcmeehan) [inline presto-iceberg/src/test/java/com/facebook/presto/iceberg/nessie/TestIcebergSmokeNessie.java:146]: There are too many tests that try to accomplish similar tasks through multi-threading.  Ultimately, these style of tests worry me because they end up being highly environment-specific--suspectible to GC pauses, noisy neighbors and such.  Can we think of a different way to accomplish this that isn't time depenendent?

**Review by tdcmeehan — CHANGES_REQUESTED** [review]

File: presto-iceberg/src/test/java/com/facebook/presto/iceberg/nessie/TestIcebergSmokeNessie.java (hunk 3 above)
Comment (pratyakshsharma) [reply presto-iceberg/src/test/java/com/facebook/presto/iceberg/nessie/TestIcebergSmokeNessie.java:146]: I am currently working on this. Lets see if I am able to come up with something.

Comment (tdcmeehan) [reply presto-iceberg/src/test/java/com/facebook/presto/iceberg/nessie/TestIcebergSmokeNessie.java:149]: I would recommend just removing the test, since this is a table property and a functionality of the core Iceberg library.  Adding a potentially flaky test to our suites is just not worth the potential problems.

Comment (pratyakshsharma) [reply presto-iceberg/src/test/java/com/facebook/presto/iceberg/nessie/TestIcebergSmokeNessie.java:149]: @tdcmeehan So with the current approach of submitting 100 queries, the test case is able to pass consistently. I am now only trying to fine tune the number of threads and other parameters, so as to reduce the test run time and not introduce flakiness overall. 
Let me know if you still think otherwise. We can probably discuss internally on this. 
cc @agrawalreetika @imjalpreet @yingsu00 @nmahadevuni

**Review by yingsu00 — COMMENTED** [review]
I see that the test is testing commitRetries + 100 queries, which doesn’t really serve the purpose of testing this fix. Let’s just remove it as @tdcmeehan suggested.

**Comment by pratyakshsharma** [comment]
So can I go ahead and squash the commits so the PR gets ready for review, or you want me to add this test instead in product-tests @yingsu00 ?

**Comment by tdcmeehan** [comment]
IMO this is fine to go without even a product test.  We're exposing a table property and making it configurable--we should test that the table property is configured correctly and that it's exported, that's the minimum we need.

**Comment by pratyakshsharma** [comment]
@yingsu00 @agrawalreetika This is good to review now.

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -45,7 +45,7 @@ jobs:
           - ":presto-tests -P ci-only-distributed-queries"
           - ":presto-tests -P ci-only-aggregation-queries"
           - ":presto-tests -P ci-only-plan-determinism"
-          - ":presto-tests -P ci-only-resource-manager" 
+          - ":presto-tests -P ci-only-resource-manager"
```
Comment (agrawalreetika) [inline .github/workflows/tests.yml:48]: Change is not related?

**Review by agrawalreetika — APPROVED** [review]

**Comment by steveburnett** [comment]
Consider changing the release note entry in the Description. Following the [release note guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines), would something like this be a better description of this PR? 

```
== RELEASE NOTES ==

Iceberg Connector Changes
* Improve the commit.retry.num-retries table property to make configurable the number of attempts to make in case of concurrent upserts.
```

File: .github/workflows/tests.yml (hunk 4 above)
Comment (yingsu00) [reply .github/workflows/tests.yml:48]: @pratyakshsharma Can you please remove this change? It seems unrelated.

**Review by yingsu00 — APPROVED** [review]
@tdcmeehan The test was removed. Do you want to review this PR again? Thanks!

**Comment by pratyakshsharma** [comment]
@steveburnett 

So we are not improving the support, rather adding it from scratch in presto-iceberg connector. Please let me know in case of any more queries.

**Comment by pratyakshsharma** [comment]
@yingsu00 This is ready for another pass

**Comment by steveburnett** [comment]
> @steveburnett
> 
> So we are not improving the support, rather adding it from scratch in presto-iceberg connector. Please let me know in case of any more queries.

Thanks for the explanation! The [release note guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines#order-of-changes) Order of Changes section suggests "Additions: line starts with Add...". Maybe this suggestion works better?

```
== RELEASE NOTES ==

Iceberg Connector Changes
* Add the commit.retry.num-retries table property to make configurable the number of attempts to make in case of concurrent upserts.
```

**Comment by tdcmeehan** [comment]
Agreed with @steveburnett, this is a major change that users of Iceberg connector will be interested in and may want to use.  We should add a release note for it.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableProperties.java [hunk 5]
```diff
@@ -35,6 +37,7 @@ public class IcebergTableProperties
     public static final String PARTITIONING_PROPERTY = "partitioning";
     public static final String LOCATION_PROPERTY = "location";
     public static final String FORMAT_VERSION = "format_version";
+    public static final String CONCURRENCY_RETRIES = "commit_retries";
```
Comment (tdcmeehan) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableProperties.java:40]: Why name the constant `CONCURRENCY_RETRIES` when the property is `commit_retries`?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableProperties.java [hunk 6]
```diff
@@ -73,6 +76,11 @@ public IcebergTableProperties(IcebergConfig icebergConfig)
                         "Format version for the table",
                         null,
                         false))
+                .add(integerProperty(
+                        CONCURRENCY_RETRIES,
+                        "Determines the number of attempts in case of concurrent upserts",
```
Comment (tdcmeehan) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableProperties.java:81]: Concurrent updates

Comment (pratyakshsharma) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableProperties.java:81]: I think it is valid in case of inserts as well. Ultimately this boils down to some writer finding the metadata file updated, as compared to what it saw in the beginning, when it tries to commit.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableProperties.java (hunk 5 above)
Comment (pratyakshsharma) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableProperties.java:40]: Will change

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableProperties.java (hunk 6 above)
Comment (tdcmeehan) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableProperties.java:81]: By update, I am trying to convey a more general meaning that upset, which may include deletes. Feel free to use any language that encapsulates this, but I feel upsert does not.

**Comment by pratyakshsharma** [comment]
Updated release notes @steveburnett @tdcmeehan

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableProperties.java (hunk 6 above)
Comment (pratyakshsharma) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableProperties.java:81]: I can probably make it upserts and deletes then.

**Comment by steveburnett** [comment]
> Updated release notes @steveburnett @tdcmeehan

@pratyakshsharma, it looks great! Thank you.

**Comment by pratyakshsharma** [comment]
The singlestore test failure is unrelated. Raised an issue here - https://github.com/prestodb/presto/issues/21892

**Comment by pratyakshsharma** [comment]
This is good for final pass @yingsu00

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableProperties.java [hunk 7]
```diff
@@ -113,4 +121,9 @@ public static String getFormatVersion(Map<String, Object> tableProperties)
     {
         return (String) tableProperties.get(FORMAT_VERSION);
     }
+
+    public static Integer getConcurrencyRetries(Map<String, Object> tableProperties)
```
Comment (tdcmeehan) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableProperties.java:125]: Here too?

**Review by tdcmeehan — APPROVED** [review]

**Comment by pratyakshsharma** [comment]
@yingsu00 this is good to merge.
